File: src/kt/PhantasyIslandKotlinRemoteControl/PhantasyIslandKotlinRemoteControl/radio/TypeDef.kt
package PhantasyIslandKotlinRemoteControl.radio

import org.json.JSONObject
import org.json.JSONArray

/**
 * 三维坐标，单位米，格式 (x, y, z)
 */
data class XYZ(val x: Double, val y: Double, val z: Double) {
    fun toList(): List<Double> = listOf(x, y, z)
    fun toJSONArray(): JSONArray = JSONArray(toList())
    
    companion object {
        fun fromJSONArray(arr: JSONArray): XYZ = XYZ(arr.getDouble(0), arr.getDouble(1), arr.getDouble(2))
        fun fromList(list: List<Any>): XYZ = XYZ((list[0] as Number).toDouble(), (list[1] as Number).toDouble(), (list[2] as Number).toDouble())
    }
}

/**
 * 链路可达性检查的可选参数。
 */
data class RadioCheckOptions(
    var frequencyMHz: Double? = null,
    var txPowerDbm: Double? = null,
    var rxSensitivityDbm: Double? = null,
    var fresnelZoneRatio: Double? = null,
    var skipFresnelZoneCheck: Boolean? = null,
    var enableMultipath: Boolean? = null,
    var maxReflectionPaths: Double? = null,
    var maxReflectionPathLengthRatio: Double? = null,
    var defaultTxAntennaGain_dBi: Double? = null,
    var defaultRxAntennaGain_dBi: Double? = null,
    var enableAntennaPattern: Boolean? = null,
    var enableMutualCoupling: Boolean? = null,
    var couplingNegligibleThresholdWavelengths: Double? = null,
    var enableSINR: Boolean? = null,
    var receiverBandwidthHz: Double? = null,
    var minSINR_dB: Double? = null,
    var enableFrequencyIsolation: Boolean? = null,
    var enableNearFieldCorrection: Boolean? = null,
    var enableNodeBodyOcclusion: Boolean? = null
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        frequencyMHz?.let { json.put("frequencyMHz", it) }
        txPowerDbm?.let { json.put("txPowerDbm", it) }
        rxSensitivityDbm?.let { json.put("rxSensitivityDbm", it) }
        fresnelZoneRatio?.let { json.put("fresnelZoneRatio", it) }
        skipFresnelZoneCheck?.let { json.put("skipFresnelZoneCheck", it) }
        enableMultipath?.let { json.put("enableMultipath", it) }
        maxReflectionPaths?.let { json.put("maxReflectionPaths", it) }
        maxReflectionPathLengthRatio?.let { json.put("maxReflectionPathLengthRatio", it) }
        defaultTxAntennaGain_dBi?.let { json.put("defaultTxAntennaGain_dBi", it) }
        defaultRxAntennaGain_dBi?.let { json.put("defaultRxAntennaGain_dBi", it) }
        enableAntennaPattern?.let { json.put("enableAntennaPattern", it) }
        enableMutualCoupling?.let { json.put("enableMutualCoupling", it) }
        couplingNegligibleThresholdWavelengths?.let { json.put("couplingNegligibleThresholdWavelengths", it) }
        enableSINR?.let { json.put("enableSINR", it) }
        receiverBandwidthHz?.let { json.put("receiverBandwidthHz", it) }
        minSINR_dB?.let { json.put("minSINR_dB", it) }
        enableFrequencyIsolation?.let { json.put("enableFrequencyIsolation", it) }
        enableNearFieldCorrection?.let { json.put("enableNearFieldCorrection", it) }
        enableNodeBodyOcclusion?.let { json.put("enableNodeBodyOcclusion", it) }
        return json
    }

    companion object {
        fun fromJSONObject(json: JSONObject): RadioCheckOptions {
            val opt = RadioCheckOptions()
            if (json.has("frequencyMHz")) opt.frequencyMHz = json.getDouble("frequencyMHz")
            if (json.has("txPowerDbm")) opt.txPowerDbm = json.getDouble("txPowerDbm")
            if (json.has("rxSensitivityDbm")) opt.rxSensitivityDbm = json.getDouble("rxSensitivityDbm")
            if (json.has("fresnelZoneRatio")) opt.fresnelZoneRatio = json.getDouble("fresnelZoneRatio")
            if (json.has("skipFresnelZoneCheck")) opt.skipFresnelZoneCheck = json.getBoolean("skipFresnelZoneCheck")
            if (json.has("enableMultipath")) opt.enableMultipath = json.getBoolean("enableMultipath")
            if (json.has("maxReflectionPaths")) opt.maxReflectionPaths = json.getDouble("maxReflectionPaths")
            if (json.has("maxReflectionPathLengthRatio")) opt.maxReflectionPathLengthRatio = json.getDouble("maxReflectionPathLengthRatio")
            if (json.has("defaultTxAntennaGain_dBi")) opt.defaultTxAntennaGain_dBi = json.getDouble("defaultTxAntennaGain_dBi")
            if (json.has("defaultRxAntennaGain_dBi")) opt.defaultRxAntennaGain_dBi = json.getDouble("defaultRxAntennaGain_dBi")
            if (json.has("enableAntennaPattern")) opt.enableAntennaPattern = json.getBoolean("enableAntennaPattern")
            if (json.has("enableMutualCoupling")) opt.enableMutualCoupling = json.getBoolean("enableMutualCoupling")
            if (json.has("couplingNegligibleThresholdWavelengths")) opt.couplingNegligibleThresholdWavelengths = json.getDouble("couplingNegligibleThresholdWavelengths")
            if (json.has("enableSINR")) opt.enableSINR = json.getBoolean("enableSINR")
            if (json.has("receiverBandwidthHz")) opt.receiverBandwidthHz = json.getDouble("receiverBandwidthHz")
            if (json.has("minSINR_dB")) opt.minSINR_dB = json.getDouble("minSINR_dB")
            if (json.has("enableFrequencyIsolation")) opt.enableFrequencyIsolation = json.getBoolean("enableFrequencyIsolation")
            if (json.has("enableNearFieldCorrection")) opt.enableNearFieldCorrection = json.getBoolean("enableNearFieldCorrection")
            if (json.has("enableNodeBodyOcclusion")) opt.enableNodeBodyOcclusion = json.getBoolean("enableNodeBodyOcclusion")
            return opt
        }
    }
}

/**
 * 链路可达性检查请求消息。
 */
data class CheckReachabilityRequest(
    val aTx: XYZ,
    val bRx: XYZ,
    val options: RadioCheckOptions? = null
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        json.put("aTx", aTx.toJSONArray())
        json.put("bRx", bRx.toJSONArray())
        options?.let { json.put("options", it.toJSONObject()) }
        return json
    }

    companion object {
        fun fromJSONObject(json: JSONObject): CheckReachabilityRequest {
            return CheckReachabilityRequest(
                aTx = XYZ.fromJSONArray(json.getJSONArray("aTx")),
                bRx = XYZ.fromJSONArray(json.getJSONArray("bRx")),
                options = if (json.has("options")) RadioCheckOptions.fromJSONObject(json.getJSONObject("options")) else null
            )
        }
    }
}

/**
 * 更新对象位置的消息。
 */
data class UpdateObjectPosRequest(
    val objectId: String,
    val position: XYZ
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        json.put("objectId", objectId)
        json.put("position", position.toJSONArray())
        return json
    }

    companion object {
        fun fromJSONObject(json: JSONObject): UpdateObjectPosRequest {
            return UpdateObjectPosRequest(
                objectId = json.getString("objectId"),
                position = XYZ.fromJSONArray(json.getJSONArray("position"))
            )
        }
    }
}

/**
 * 更新对象的电磁属性。
 */
data class UpdateMeshRadioMaterialRequest(
    val meshId: String,
    var materialId: String? = null,
    var thickness_m: Double? = null
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        json.put("meshId", meshId)
        materialId?.let { json.put("materialId", it) }
        thickness_m?.let { json.put("thickness_m", it) }
        return json
    }

    companion object {
        fun fromJSONObject(json: JSONObject): UpdateMeshRadioMaterialRequest {
            val req = UpdateMeshRadioMaterialRequest(json.getString("meshId"))
            if (json.has("materialId")) req.materialId = json.getString("materialId")
            if (json.has("thickness_m")) req.thickness_m = json.getDouble("thickness_m")
            return req
        }
    }
}

/**
 * 电磁材料属性定义。
 */
data class RadioMaterialProperties(
    val id: String,
    val displayName: String,
    val penetrationLoss_dBPerMeter: Double,
    val reflectionCoefficient: Double,
    val defaultThickness_m: Double
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        json.put("id", id)
        json.put("displayName", displayName)
        json.put("penetrationLoss_dBPerMeter", penetrationLoss_dBPerMeter)
        json.put("reflectionCoefficient", reflectionCoefficient)
        json.put("defaultThickness_m", defaultThickness_m)
        return json
    }

    companion object {
        fun fromJSONObject(json: JSONObject): RadioMaterialProperties {
            return RadioMaterialProperties(
                id = json.getString("id"),
                displayName = json.getString("displayName"),
                penetrationLoss_dBPerMeter = json.getDouble("penetrationLoss_dBPerMeter"),
                reflectionCoefficient = json.getDouble("reflectionCoefficient"),
                defaultThickness_m = json.getDouble("defaultThickness_m")
            )
        }
    }
}

/**
 * 摇杆输入消息。
 */
data class JoyStickInput(
    val vx: Double = 0.0,
    val vy: Double = 0.0,
    val vz: Double = 0.0,
    val yawRate: Double = 0.0
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        json.put("vx", vx)
        json.put("vy", vy)
        json.put("vz", vz)
        json.put("yawRate", yawRate)
        return json
    }

    companion object {
        fun fromJSONObject(json: JSONObject): JoyStickInput {
            return JoyStickInput(
                vx = json.optDouble("vx", 0.0),
                vy = json.optDouble("vy", 0.0),
                vz = json.optDouble("vz", 0.0),
                yawRate = json.optDouble("yawRate", 0.0)
            )
        }
    }
}

File: src/ts/src/radio_py/typeDef.ts
/**
 * TypeScript 类型定义。
 */

// ---------------------------------------------------------------------------
// XYZ — 三维坐标，单位米，格式 [x, y, z]
// ---------------------------------------------------------------------------
export type XYZ = [number, number, number];

/** 链路可达性检查的可选参数。所有字段均为可选，服务端会使用各自的默认值。 */
export interface RadioCheckOptions {
    /** 信号频率（MHz），默认 2400（2.4GHz，无人机常用频段） */
    frequencyMHz?: number;
    /** 发射功率（dBm），默认 20 */
    txPowerDbm?: number;
    /** 接收灵敏度（dBm），默认 -90。接收功率低于此值视为不可达 */
    rxSensitivityDbm?: number;
    /** 菲涅尔区清晰度阈值（0~1），默认 0.6。第一菲涅尔区 60% 以上无遮挡时，信号衰减可忽略不计 */
    fresnelZoneRatio?: number;
    /** 是否跳过菲涅尔区体积检测（默认 false）。设为 True 时仅做中心射线 LOS 检测，显著提升性能 */
    skipFresnelZoneCheck?: boolean;
    /** 是否启用多径反射计算（默认 false）。计算开销约为当前查询的 2~4 倍 */
    enableMultipath?: boolean;
    /** 最大反射路径数（默认 8），仅保留最强的 N 条。用于控制多径计算开销 */
    maxReflectionPaths?: number;
    /** 反射路径最大长度与直射距离的比值（默认 2.0） */
    maxReflectionPathLengthRatio?: number;
    /** 发射端默认天线增益（dBi），默认 0（各向同性天线）。在非多节点网络扩展模式下应用于链路预算计算 */
    defaultTxAntennaGain_dBi?: number;
    /** 接收端默认天线增益（dBi），默认 0（各向同性天线）。在非多节点网络扩展模式下应用于链路预算计算 */
    defaultRxAntennaGain_dBi?: number;
    /** 是否启用天线方向图增益修正（默认 false）。启用后根据天线指向方向查询增益，替代默认的各向同性假设 */
    enableAntennaPattern?: boolean;
    /** 是否启用多天线互耦合修正（默认 false）。当天线间距小于阈值时，天线之间的感应电流会导致额外的发射功率损耗 */
    enableMutualCoupling?: boolean;
    /** 互耦合可忽略的天线间距阈值（单位：波长数），默认 10。当两天线间距 d / λ ≥ 此阈值时，互耦合损耗视为 0 */
    couplingNegligibleThresholdWavelengths?: number;
    /** 是否启用同频 SINR（信噪干扰比）计算（默认 false）。计算复杂度为 O(N²) */
    enableSINR?: boolean;
    /** 接收机噪声带宽（Hz），默认 20e6（20 MHz，802.11n 标准信道）。用于计算热噪声底 */
    receiverBandwidthHz?: number;
    /** SINR 最低阈值（dB），默认 10。SINR > minSINR_dB 时链路质量合格 */
    minSINR_dB?: number;
    /** 是否启用频段隔离计算（默认 false）。可大幅减少多频段部署时的干扰矩阵计算量 */
    enableFrequencyIsolation?: boolean;
    /** 是否启用近场修正（默认 false）。当两点距离 < 5λ 时启用，防止 FSPL 模型出现负值 */
    enableNearFieldCorrection?: boolean;
    /** 保留字段，暂未使用。用于检测节点物理体之间的相互遮挡 */
    enableNodeBodyOcclusion?: boolean;
}

/** 链路可达性检查请求消息。 */
export interface CheckReachabilityRequest {
    /** 发射端的三维坐标 */
    aTx: XYZ;
    /** 接收端的三维坐标 */
    bRx: XYZ;
    /** 可选的链路检查参数 */
    options?: RadioCheckOptions;
}

/** 更新对象位置的消息。 */
export interface UpdateObjectPosRequest {
    /** 对象的唯一标识 */
    objectId: string;
    /** 新的坐标 */
    position: XYZ;
}

/** 更新对象的电磁属性。 */
export interface UpdateMeshRadioMaterialRequest {
    /** 网格的唯一标识 */
    meshId: string;
    /** 材料类型 Id，如 "concrete", "wood", "metal" 等。见 getAllRadioMaterial 接口返回值 */
    materialId?: string;
    /** 单面物体情况下的物体默认厚度，单位米 */
    thickness_m?: number;
}

/** 电磁材料属性定义。 */
export interface RadioMaterialProperties {
    /** 材料的唯一标识，如 "concrete", "wood", "metal" 等 */
    id: string;
    /** 材料的显示名称 */
    displayName: string;
    /** 每米的穿透损耗，单位 dB/m */
    penetrationLoss_dBPerMeter: number;
    /** 反射系数，范围 0~1，表示入射信号被反射回去的比例 */
    reflectionCoefficient: number;
    /** 默认厚度，单位米。对于单面物体（如墙壁）来说，这个值用于估算穿透损耗；对于双面物体（如地面）来说，这个值可以忽略 */
    defaultThickness_m: number;
}

/** 摇杆输入消息。 */
export interface JoyStickInput {
    /** 前后速度输入，范围 -1~1，正值表示向前 */
    vx: number;
    /** 左右速度输入，范围 -1~1，正值表示向右 */
    vy: number;
    /** 上下速度输入，范围 -1~1，正值表示向上 */
    vz: number;
    /** 偏航角速度输入，范围 -1~1，正值表示顺时针旋转 */
    yawRate: number;
}
